File: src/lib/api/api-channels.ts
import { tauriInvoke } from "$lib/api/client";

export type ApiChannel = {
  id: string;
  name: string;
  channel_type: string;
  base_url: string;
  auth_type: string;
  api_key: string | null;
  models_endpoint: string | null;
  chat_endpoint: string | null;
  stream_endpoint: string | null;
  models_mode: string;
  enabled: boolean;
  sort_order: number;
  config_json: Record<string, unknown>;
  created_at: number;
  updated_at: number;
};

export type ApiChannelModel = {
  id: string;
  channel_id: string;
  model_id: string;
  display_name: string | null;
  model_type: string | null;
  context_window: number | null;
  max_output_tokens: number | null;
  capabilities_json: Record<string, unknown>;
  pricing_json: Record<string, unknown>;
  default_parameters_json: Record<string, unknown>;
  sort_order: number;
  config_json: Record<string, unknown>;
};

export type CreateApiChannelInput = {
  name: string;
  channel_type: string;
  base_url: string;
  auth_type: string;
  api_key: string | null;
  models_endpoint: string | null;
  chat_endpoint: string | null;
  stream_endpoint: string | null;
  models_mode: string;
  enabled: boolean;
  sort_order: number;
  config_json: Record<string, unknown>;
};

export type UpdateApiChannelInput = CreateApiChannelInput;

export type UpsertApiChannelModelInput = {
  channel_id: string;
  model_id: string;
  display_name: string | null;
  model_type: string | null;
  context_window: number | null;
  max_output_tokens: number | null;
  capabilities_json: Record<string, unknown>;
  pricing_json: Record<string, unknown>;
  default_parameters_json: Record<string, unknown>;
  sort_order: number;
  config_json: Record<string, unknown>;
};

export type ApiChannelTestResponse = {
  model_id: string;
  response_text: string;
};

export function listApiChannels() {
  return tauriInvoke<ApiChannel[]>("list_api_channels");
}

export function getApiChannel(id: string) {
  return tauriInvoke<ApiChannel>("get_api_channel", { id });
}

export function listApiChannelModels(channelId: string) {
  return tauriInvoke<ApiChannelModel[]>("list_api_channel_models", { channelId });
}

export function fetchApiChannelRemoteModels(channelId: string) {
  return tauriInvoke<ApiChannelModel[]>("fetch_api_channel_remote_models", { channelId });
}

export function createApiChannel(input: CreateApiChannelInput) {
  return tauriInvoke<ApiChannel>("create_api_channel", { input });
}

export function updateApiChannel(id: string, input: UpdateApiChannelInput) {
  return tauriInvoke<ApiChannel>("update_api_channel", { id, input });
}

export function deleteApiChannel(id: string) {
  return tauriInvoke<void>("delete_api_channel", { id });
}

export function upsertApiChannelModel(input: UpsertApiChannelModelInput) {
  return tauriInvoke<ApiChannelModel>("upsert_api_channel_model", { input });
}

export function deleteApiChannelModel(channelId: string, modelId: string) {
  return tauriInvoke<void>("delete_api_channel_model", { channelId, modelId });
}

export function refreshApiChannelModels(channelId: string) {
  return tauriInvoke<ApiChannelModel[]>("refresh_api_channel_models", { channelId });
}

export function testApiChannelMessage(channelId: string, modelId: string) {
  return tauriInvoke<ApiChannelTestResponse>("test_api_channel_message", { channelId, modelId });
}

File: src-tauri/src/domain/api_channels.rs
use serde::{Deserialize, Serialize};

use crate::domain::common::{Id, TimestampMs};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiChannel {
    pub id: Id,
    pub name: String,
    pub channel_type: String,
    pub base_url: String,
    pub auth_type: String,
    pub api_key: Option<String>,
    pub models_endpoint: Option<String>,
    pub chat_endpoint: Option<String>,
    pub stream_endpoint: Option<String>,
    pub models_mode: String,
    pub enabled: bool,
    pub sort_order: i64,
    pub config_json: serde_json::Value,
    pub created_at: TimestampMs,
    pub updated_at: TimestampMs,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiChannelModel {
    pub id: Id,
    pub channel_id: Id,
    pub model_id: String,
    pub display_name: Option<String>,
    pub model_type: Option<String>,
    pub context_window: Option<i64>,
    pub max_output_tokens: Option<i64>,
    pub capabilities_json: serde_json::Value,
    pub pricing_json: serde_json::Value,
    pub default_parameters_json: serde_json::Value,
    pub sort_order: i64,
    pub config_json: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateApiChannelInput {
    pub name: String,
    pub channel_type: String,
    pub base_url: String,
    pub auth_type: String,
    pub api_key: Option<String>,
    pub models_endpoint: Option<String>,
    pub chat_endpoint: Option<String>,
    pub stream_endpoint: Option<String>,
    pub models_mode: String,
    pub enabled: bool,
    pub sort_order: i64,
    pub config_json: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateApiChannelInput {
    pub name: String,
    pub channel_type: String,
    pub base_url: String,
    pub auth_type: String,
    pub api_key: Option<String>,
    pub models_endpoint: Option<String>,
    pub chat_endpoint: Option<String>,
    pub stream_endpoint: Option<String>,
    pub models_mode: String,
    pub enabled: bool,
    pub sort_order: i64,
    pub config_json: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpsertApiChannelModelInput {
    pub channel_id: Id,
    pub model_id: String,
    pub display_name: Option<String>,
    pub model_type: Option<String>,
    pub context_window: Option<i64>,
    pub max_output_tokens: Option<i64>,
    pub capabilities_json: serde_json::Value,
    pub pricing_json: serde_json::Value,
    pub default_parameters_json: serde_json::Value,
    pub sort_order: i64,
    pub config_json: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiChannelTestResponse {
    pub model_id: String,
    pub response_text: String,
}
